File: core/src/main/java/dev/vibris/core/CoreProbe.java
package dev.vibris.core;

import dev.vibris.protocol.v1.SceneContext;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class CoreProbe {
    private static final int TRACE_LIMIT = 1_024;
    private static final int HISTORY_LIMIT = 128;

    private final List<String> executionOrder = new ArrayList<>();
    private final List<String> executionEvents = new ArrayList<>();
    private final List<SceneContext> contextSnapshots = new ArrayList<>();
    private final List<JobTrace> jobTraces = new ArrayList<>();
    private final List<SourceTransition> sourceTransitions = new ArrayList<>();
    private final Map<String, List<String>> sourceStates = new LinkedHashMap<>();
    private final Map<String, Integer> executionCounts = new LinkedHashMap<>();
    private int concurrentJobs;
    private int maxConcurrentJobs;
    private int requestRegistrySize;
    private int requestRegistryPeak;
    private int sourceRegistrySize;
    private int sourceRegistryPeak;
    private int queueSize;
    private int queuePeak;

    synchronized void jobStarted(String requestId) {
        boundedAdd(executionOrder, requestId, TRACE_LIMIT);
        executionCounts.merge(requestId, 1, Integer::sum);
        trimMap(executionCounts);
        concurrentJobs++;
        maxConcurrentJobs = Math.max(maxConcurrentJobs, concurrentJobs);
    }

    synchronized void contextApplied(String requestId, String workspaceId, SceneContext context) {
        boundedAdd(contextSnapshots, context, TRACE_LIMIT);
        boundedAdd(jobTraces, new JobTrace(requestId, workspaceId, context), TRACE_LIMIT);
    }

    synchronized void event(String requestId, String event) {
        boundedAdd(executionEvents, requestId + ":" + event, TRACE_LIMIT);
    }

    synchronized void jobStopped() {
        concurrentJobs--;
    }

    synchronized void sourceTransition(String uuid, String from, String to) {
        List<String> states = sourceStates.computeIfAbsent(uuid, ignored -> new ArrayList<>());
        trimMap(sourceStates);
        if (states.isEmpty() && !from.isEmpty()) states.add(from);
        states.add(to);
        boundedAdd(sourceTransitions, new SourceTransition(uuid, from, to), TRACE_LIMIT);
    }

    synchronized void registries(int requests, int sources, int queued) {
        requestRegistrySize = requests;
        requestRegistryPeak = Math.max(requestRegistryPeak, requests);
        sourceRegistrySize = sources;
        sourceRegistryPeak = Math.max(sourceRegistryPeak, sources);
        queueSize = queued;
        queuePeak = Math.max(queuePeak, queued);
    }

    public synchronized List<String> executionOrder() {
        return List.copyOf(executionOrder);
    }

    public synchronized List<String> executionEvents() {
        return List.copyOf(executionEvents);
    }

    public synchronized List<SceneContext> contextSnapshots() {
        return List.copyOf(contextSnapshots);
    }

    public synchronized List<String> sourceStates(String uuid) {
        return List.copyOf(sourceStates.getOrDefault(uuid, List.of()));
    }

    public synchronized int maxConcurrentJobs() {
        return maxConcurrentJobs;
    }

    public synchronized int executionCount(String requestId) {
        return executionCounts.getOrDefault(requestId, 0);
    }

    public synchronized int requestRegistrySize() {
        return requestRegistrySize;
    }

    public int requestRegistryCapacity() {
        return VibrisCoreEngine.REQUEST_REGISTRY_CAPACITY;
    }

    public synchronized int sourceRegistrySize() {
        return sourceRegistrySize;
    }

    public int sourceRegistryCapacity() {
        return SourceRegistry.CAPACITY;
    }

    public synchronized int queueSize() {
        return queueSize;
    }

    public int queueCapacity() {
        return FairJobScheduler.CAPACITY;
    }

    public synchronized Snapshot snapshot() {
        return new Snapshot(List.copyOf(jobTraces), List.copyOf(sourceTransitions), List.copyOf(executionOrder),
            List.copyOf(executionEvents), Map.copyOf(executionCounts), maxConcurrentJobs, queuePeak,
            requestRegistrySize, requestRegistryPeak, sourceRegistrySize, sourceRegistryPeak);
    }

    private static <T> void boundedAdd(List<T> values, T value, int limit) {
        if (values.size() == limit) values.removeFirst();
        values.add(value);
    }

    private static void trimMap(Map<?, ?> values) {
        while (values.size() > HISTORY_LIMIT) values.remove(values.keySet().iterator().next());
    }

    public record JobTrace(String requestId, String workspaceId, SceneContext context) {
    }

    public record SourceTransition(String uuid, String from, String to) {
    }

    public record Snapshot(List<JobTrace> jobTraces, List<SourceTransition> sourceTransitions,
                           List<String> executionOrder, List<String> executionEvents,
                           Map<String, Integer> executionCounts,
                           int maxConcurrentJobs, int queuePeak, int requestRegistrySize,
                           int requestRegistryPeak, int sourceRegistrySize, int sourceRegistryPeak) {
    }
}

File: core/src/main/kotlin/dev/vibris/core/CoreProbe.kt
package dev.vibris.core

import dev.vibris.protocol.v2.SceneContext
import java.util.ArrayList
import java.util.LinkedHashMap

class CoreProbe {
    private val executionOrder = ArrayList<String>()
    private val executionEvents = ArrayList<String>()
    private val contextSnapshots = ArrayList<SceneContext>()
    private val jobTraces = ArrayList<JobTrace>()
    private val sourceTransitions = ArrayList<SourceTransition>()
    private val sourceStates = LinkedHashMap<String, MutableList<String>>()
    private val executionCounts = LinkedHashMap<String, Int>()
    private var concurrentJobs = 0
    private var maxConcurrentJobs = 0
    private var requestRegistrySize = 0
    private var requestRegistryPeak = 0
    private var sourceRegistrySize = 0
    private var sourceRegistryPeak = 0
    private var queueSize = 0
    private var queuePeak = 0

    @Synchronized
    fun jobStarted(requestId: String) {
        boundedAdd(executionOrder, requestId, TRACE_LIMIT)
        executionCounts.merge(requestId, 1) { left, right -> left + right }
        trimMap(executionCounts)
        concurrentJobs++
        maxConcurrentJobs = maxOf(maxConcurrentJobs, concurrentJobs)
    }

    @Synchronized
    fun contextApplied(requestId: String, workspaceId: String, context: SceneContext) {
        boundedAdd(contextSnapshots, context, TRACE_LIMIT)
        boundedAdd(jobTraces, JobTrace(requestId, workspaceId, context), TRACE_LIMIT)
    }

    @Synchronized
    fun event(requestId: String, event: String) {
        boundedAdd(executionEvents, "$requestId:$event", TRACE_LIMIT)
    }

    @Synchronized
    fun jobStopped() {
        concurrentJobs--
    }

    @Synchronized
    fun sourceTransition(uuid: String, from: String, to: String) {
        val states = sourceStates.computeIfAbsent(uuid) { ArrayList() }
        trimMap(sourceStates)
        if (states.isEmpty() && from.isNotEmpty()) {
            states.add(from)
        }
        states.add(to)
        boundedAdd(sourceTransitions, SourceTransition(uuid, from, to), TRACE_LIMIT)
    }

    @Synchronized
    fun registries(requests: Int, sources: Int, queued: Int) {
        requestRegistrySize = requests
        requestRegistryPeak = maxOf(requestRegistryPeak, requests)
        sourceRegistrySize = sources
        sourceRegistryPeak = maxOf(sourceRegistryPeak, sources)
        queueSize = queued
        queuePeak = maxOf(queuePeak, queued)
    }

    @Synchronized
    fun executionOrder(): List<String> = java.util.List.copyOf(executionOrder)

    @Synchronized
    fun executionEvents(): List<String> = java.util.List.copyOf(executionEvents)

    @Synchronized
    fun contextSnapshots(): List<SceneContext> = java.util.List.copyOf(contextSnapshots)

    @Synchronized
    fun sourceStates(uuid: String): List<String> =
        java.util.List.copyOf(sourceStates.getOrDefault(uuid, emptyList()))

    @Synchronized
    fun maxConcurrentJobs(): Int = maxConcurrentJobs

    @Synchronized
    fun executionCount(requestId: String): Int = executionCounts.getOrDefault(requestId, 0)

    @Synchronized
    fun requestRegistrySize(): Int = requestRegistrySize

    fun requestRegistryCapacity(): Int = VibrisCoreEngine.REQUEST_REGISTRY_CAPACITY

    @Synchronized
    fun sourceRegistrySize(): Int = sourceRegistrySize

    fun sourceRegistryCapacity(): Int = SourceRegistry.CAPACITY

    @Synchronized
    fun queueSize(): Int = queueSize

    fun queueCapacity(): Int = FairJobScheduler.CAPACITY

    @Synchronized
    fun snapshot(): Snapshot =
        Snapshot(
            java.util.List.copyOf(jobTraces),
            java.util.List.copyOf(sourceTransitions),
            java.util.List.copyOf(executionOrder),
            java.util.List.copyOf(executionEvents),
            java.util.Map.copyOf(executionCounts),
            maxConcurrentJobs,
            queuePeak,
            requestRegistrySize,
            requestRegistryPeak,
            sourceRegistrySize,
            sourceRegistryPeak,
        )

    @JvmRecord
    data class JobTrace(
        val requestId: String,
        val workspaceId: String,
        val context: SceneContext,
    )

    @JvmRecord
    data class SourceTransition(
        val uuid: String,
        val from: String,
        val to: String,
    )

    @JvmRecord
    data class Snapshot(
        val jobTraces: List<JobTrace>,
        val sourceTransitions: List<SourceTransition>,
        val executionOrder: List<String>,
        val executionEvents: List<String>,
        val executionCounts: Map<String, Int>,
        val maxConcurrentJobs: Int,
        val queuePeak: Int,
        val requestRegistrySize: Int,
        val requestRegistryPeak: Int,
        val sourceRegistrySize: Int,
        val sourceRegistryPeak: Int,
    )

    companion object {
        private const val TRACE_LIMIT = 1_024
        private const val HISTORY_LIMIT = 128

        private fun <T> boundedAdd(values: MutableList<T>, value: T, limit: Int) {
            if (values.size == limit) {
                values.removeAt(0)
            }
            values.add(value)
        }

        private fun trimMap(values: MutableMap<*, *>) {
            while (values.size > HISTORY_LIMIT) {
                values.remove(values.keys.iterator().next())
            }
        }
    }
}